cfme/common/vm_console.py

```diff
@@ -0,0 +1,179 @@
+        self.name = name
+        self.selenium = selenium
+        self.console_handle = console_handle
+        self.appliance_handle = appliance_handle
+        self.provider = provider
+
+    ###
+    # Methods
+    #
+    def _get_openstack_banner(self):
+        text = self.selenium.find_element_by_id(banner_id).text
+        return text
+
+    def _get_vmware_rhev_connection_status_text(self):
+        text = self.selenium.find_element_by_xpath(connection_status_xpath).text
+        return text
+
+    def get_banner(self):
+        """Get the text of the banner above the console screen."""
+        self.switch_to_console()
+        if self.provider.one_of(OpenStackProvider):
+            text = self._get_openstack_banner()
+        else:
+            text = self._get_vmware_rhev_connection_status_text()
+        logger.info('Read following text from console banner: {}'.format(text))
+        self.switch_to_appliance()
+        return text
+
+    def _get_openstack_canvas(self):
+        return self.selenium.find_element_by_id(canvas_id)
+
+    def _get_vmware_rhev_canvas(self):
+        return self.selenium.find_element_by_xpath(canvas_xapth)
+
+    def get_screen(self):
+        """
+        Retrieve the bit map from the canvas widget that represents the console screen.
+
+        Returns it as a binary string.
+
+        Implementation:
+        The canvas tag has a method toDataURL() which one can use in javascript to
+        obtain the canvas image  base64 encoded.   Examples of how to do this can be
+        seen here:
+
+            https://qxf2.com/blog/selenium-html5-canvas-verify-what-was-drawn/
+            https://stackoverflow.com/questions/38316402/how-to-save-a-canvas-as-png-in-selenium
+        """
+        self.switch_to_console()
+
+        # Get the canvas element
+        canvas = self.get_canvas()
+
+        # Now run some java script to get the contents of the canvas element
+        # base 64 encoded.
+        image_base64_url = self.selenium.execute_script(
+            "return arguments[0].toDataURL('image/jpeg',1);",
+            canvas
+        )
+
+        # The results will look like:
+        #
+        #   data:image/jpeg;base64,iVBORw0KGgoAAAANSUhEUgAAAfQAAABkCAYAAABwx8J9AA...
+        #
+        # So parse out the data from the non image data from the URL:
+        image_base64 = image_base64_url.split(",")[1]
+
+        # Now convert to binary:
+        image_jpeg = base64.b64decode(image_base64)
+
+        self.switch_to_appliance()
+        return image_jpeg
+
+    def get_screen_text(self):
+        """
+        Return the text from a text console.
+
+        Uses OCR to scrape the text from the console image taken at the time of the call.
+        """
+        image_str = self.get_screen()
+
+        # Write the image string to a file as pytesseract requires
+        # a file, and doesn't take a string.
+        tmp_file = tempfile.NamedTemporaryFile(suffix='.jpeg')
```

Comment:
**REQUIRED:** We should rather do this with an in-memory object, rather than writing out to a file.

https://docs.python.org/release/2.7/library/io.html#binary-i-o